infrastructure/containerd/image_service_test.go

```diff
@@ -2,133 +2,669 @@ package containerd_test
-	. "github.com/onsi/gomega"
-
+	"github.com/golang/mock/gomock"
+	g "github.com/onsi/gomega"
+	"github.com/opencontainers/go-digest"
 	"github.com/weaveworks/flintlock/core/models"
 	"github.com/weaveworks/flintlock/core/ports"
 	"github.com/weaveworks/flintlock/infrastructure/containerd"
+	"github.com/weaveworks/flintlock/infrastructure/mock"
 )
 
 const (
-	testImageVolume    = "docker.io/library/alpine:3.14.1"
-	testImageKernel    = "docker.io/linuxkit/kernel:5.4.129"
-	testSnapshotter    = "native"
-	testOwnerNamespace = "int_ns"
-	testOwnerUsageID   = "vol1"
-	testOwnerName      = "imageservice-get-test"
-	testContainerdNs   = "flintlock_test_ctr"
+	testImage   = "testimage"
+	testOwner   = "testowner"
+	testOwnerID = "testownerid"
 )
 
-func TestImageService_Integration(t *testing.T) {
-	if !runContainerDTests() {
-		t.Skip("skipping containerd image service integration test")
+// TestImageService_Pull tests a successful Pull.
+func TestImageService_Pull(t *testing.T) {
+	g.RegisterTestingT(t)
+
+	mockCtrl := gomock.NewController(t)
+	containerdClient := mock.NewMockClient(mockCtrl)
+	leasesManager := mock.NewMockManager(mockCtrl)
+	svcConfig := containerd.Config{
+		SnapshotterKernel: "native",
+		SnapshotterVolume: "devmapper",
+		SocketPath:        "/something",
+		Namespace:         "unit_test_ns",
 	}
+	ctx := context.Background()
+	client := containerd.NewImageServiceWithClient(&svcConfig, containerdClient)
+
+	leasesManager.EXPECT().
+		List(gomock.Any(), fmt.Sprintf("id==flintlock/%s", testOwner))
+	leasesManager.EXPECT().
+		Create(gomock.Any(), gomock.Any())
+	containerdClient.EXPECT().
+		LeasesService().
+		Return(leasesManager)
+	containerdClient.EXPECT().
+		Pull(gomock.Any(), testImage)
+
+	err := client.Pull(ctx, &ports.ImageSpec{
+		ImageName: testImage,
+		Owner:     testOwner,
+	})
+	g.Expect(err).ToNot(g.HaveOccurred())
+}
+
+// TestImageService_Pull_failedLease tests what happens when something goes
+// wrong with Leases.
+func TestImageService_Pull_failedLease(t *testing.T) {
+	g.RegisterTestingT(t)
+
+	mockCtrl := gomock.NewController(t)
+	containerdClient := mock.NewMockClient(mockCtrl)
+	leasesManager := mock.NewMockManager(mockCtrl)
+	svcConfig := containerd.Config{
+		SnapshotterKernel: "native",
+		SnapshotterVolume: "devmapper",
+		SocketPath:        "/something",
+		Namespace:         "unit_test_ns",
+	}
+	ctx := context.Background()
+	client := containerd.NewImageServiceWithClient(&svcConfig, containerdClient)
 
-	RegisterTestingT(t)
+	leasesManager.EXPECT().
+		List(gomock.Any(), fmt.Sprintf("id==flintlock/%s", testOwner)).
+		Return(nil, errors.New("nope"))
+	containerdClient.EXPECT().
+		LeasesService().
+		Return(leasesManager)
 
-	client, ctx := testCreateClient(t)
-	namespaceCtx := namespaces.WithNamespace(ctx, testContainerdNs)
+	err := client.Pull(ctx, &ports.ImageSpec{
+		ImageName: testImage,
+		Owner:     testOwner,
+	})
+	g.Expect(err).To(g.HaveOccurred())
```

Comment:
let's check that the error is the one we expect here (ditto all the rest too probably)

Reply:
I wouldn't go that deep as they can be very ugly by design as we just pass through the error without checking anything on them and at the end is a wrapped error wrapped into a new error that's wrapped into a new error that's wrapped ................. and who knows how deep is that. From the point we call `client.Pull` the error is basically the same containerd will throw back and we just wrap them with `fmt.Errorf(..., err)`.

Reply:
That would be cool if we can refactor this code to be less chaotic (and a bit more complex) on error handling, but with the code we have now, I don't think it has any use to test if the error is `nope` wrapped in N layers.

Reply:
you can do like a `ContainSubstring` and look for a certain piece

Reply:
but yeh LOTS of refactoring would be good in all this area

Reply:
> you can do like a `ContainSubstring` and look for a certain piece

well, it's the same if I saw "occurred" because I know it will contain what I tell it to use
```
		Return(nil, errors.New("nope"))
```
and we have zero custom error handling in this module. Actually we have one in content store, but this is the ImageService.

Reply:
don't we wrap it? if we hit the failure at that point i would expect to have substring `"listing existing containerd leases: nope"`

Reply:
yes we wrap, but we wrap all of them. so it will be always `something: something: something: something: nope` (and the amount of `something` changes based on what we called), but 100% it will contain `nope`.

Reply:
so we can still do
```suggestion
	g.Expect(err).To(g.MatchError(ContainSubstring("listing existing containerd leases: nope")))
```

Reply:
(pure personal opinion) In my eyes it falls into the "we can test it for coverage and because looks good, but useless"

Reply:
i mean it doesn;t look _great_ :trollface: 

but seriously 90% of the reviews i have done in my life have been "do we not want to catch that error?" followed by an "oh shit yeh!" response. so checking errors are returned when we expect and in the way we expect them is an unhealthy obsession of mine 😭